socrata/socrata_upload.py

```diff
@@ -0,0 +1,337 @@
+    # Socrata uploads. Automatically convert any array type columns to string.
+    columns.loc[columns["type"] == "array(varchar)", "column"] = (
+        "ARRAY_JOIN("
+        + columns[columns["type"] == "array(varchar)"]["column"]
+        + ", ', ') AS "
+        + columns[columns["type"] == "array(varchar)"]["column"]
+    )
+
+    if not years:
+        query = (
+            "SELECT\n"
+            + row_identifier
+            + "\n"
+            + ",\n".join(columns["column"])
+            + "\nFROM "
+            + athena_asset
+            + "\nLIMIT 1000"
+        )
+
+    elif years is not None and not township:
+        query = (
+            "SELECT\n"
+            + row_identifier
+            + "\n"
+            + ",\n".join(columns["column"])
+            + "\nFROM "
+            + athena_asset
+            + "\nWHERE year = %(year)s"
+            + "\nLIMIT 1000"
+        )
+
+    elif years is not None and township is not None:
+        query = (
+            "SELECT\n"
+            + row_identifier
+            + "\n"
+            + ",\n".join(columns["column"])
+            + "\nFROM "
+            + athena_asset
+            + "\nWHERE year = %(year)s"
+            + "\nAND township_code = %(township)s"
+            + "\nLIMIT 1000"
+        )
+
+    return query
+
+
+def upload(method, asset_id, sql_query, overwrite, year=None, township=None):
+    """
+    Function to perform the upload to Socrata. `puts` or `posts` depending on
+    user's choice to overwrite existing data.
+    """
+    url = (
+        "https://datacatalog.cookcountyil.gov/resource/"
+        + asset_id
+        + ".json?$$app_token="
+        + app_token
+    )
+
+    print_message = "Overwriting" if overwrite else "Updating"
+
+    if not year:
+        query_conditionals = {}
+        print_message = print_message + " all years for asset " + asset_id
+    if year is not None and not township:
+        query_conditionals = {"year": year}
+        print_message = (
+            print_message + " year: " + year + " for asset " + asset_id
+        )
+    if year is not None and township is not None:
+        query_conditionals = {"year": year, "township": township}
+        print_message = (
+            print_message
+            + " township: "
+            + township
+            + ", year: "
+            + year
+            + " for asset "
+            + asset_id
+        )
+
+    print(print_message)
+    if method == "put":
+        response = requests.put(
+            url=url,
+            data=cursor.execute(sql_query, query_conditionals)
+            .as_pandas()
+            .to_json(orient="records"),
+            auth=auth,
+        )
+
+    elif method == "post":
+        response = requests.post(
+            url=url,
+            data=cursor.execute(sql_query, query_conditionals)
+            .as_pandas()
+            .to_json(orient="records"),
+            auth=auth,
+        )
```

Comment:
**[Question, non-blocking]** I wonder if we should call [`raise_for_status()`](https://3.python-requests.org/user/quickstart/#response-status-codes) on these responses so that we can surface errors? More broadly, what's our desired behavior in case an upload fails partway through a job?